### DISCUSSION TITLE
Trouble Setting Up Prisma For Production Inside a Docker Container (Nest)

### DISCUSSION THREAD

**Discussion by dufrtss** [prisma/prisma#23383] [Q&A] [answered]
## Hi 👋🏻 ,
I'm building an app using Nest and Prisma, when running locally, my scripts for running the development version, building and starting it on production through the dist folder (I use npm), works fine, I can run both in development and production in my machine.

## The problem:
When running it on Docker, I am getting some strange behaviors and I'd like some feedback on how to fix it and if there are better approaches to this problem. My Dockerfile is a multi-step image builder, as follows:

```Dockerfile
###################
# BUILD FOR LOCAL DEVELOPMENT
###################

FROM node:21-alpine AS development

RUN apk add --no-cache bash

ENV NODE_ENV development

USER node

WORKDIR /usr/src/app

COPY --chown=node:node package*.json ./

RUN npm ci

COPY --chown=node:node . .

RUN chmod +x ./docker/wait-for-it.sh ./docker/docker-entrypoint.sh

ENTRYPOINT [ "/usr/src/app/docker/docker-entrypoint.sh" ]

###################
# BUILD FOR PRODUCTION
###################

FROM node:21-alpine AS build

ENV NODE_ENV production

USER node

WORKDIR /usr/src/app

COPY --chown=node:node package*.json ./

COPY --chown=node:node --from=development /usr/src/app/node_modules ./node_modules

COPY --chown=node:node . .

RUN ["npm", "run", "build"]

RUN npm ci --only=production --omit=dev && npm cache clean --force


###################
# PRODUCTION
###################

FROM node:21-alpine AS production

RUN apk add --no-cache bash

ENV NODE_ENV production

USER node

WORKDIR /usr/src/app

COPY --chown=node:node package*.json ./

COPY --chown=node:node --from=build /usr/src/app/node_modules ./node_modules

COPY --chown=node:node --from=build /usr/src/app/docker ./docker

COPY --chown=node:node --from=build /usr/src/app/dist ./dist

RUN chmod +x ./docker/wait-for-it.sh ./docker/docker-entrypoint.sh

ENTRYPOINT [ "/usr/src/app/docker/docker-entrypoint.sh" ]

CMD ["npm", "run", "start:prod"]

```

### Some additional info on it:

I have a make file that builds this image and calls a docker compose file, which creates the container based on the generated image and runs it if the target is the development step, otherwise Dockerfile runs it as production, as said by default with CMD on the end of the file.

I need to install bash to execute my docker-entrypoint.sh, which is responsible for checking database availability and and running prisma generate and prisma migrate deploy.

### How to reproduce (optional)

1. Run 'make deploy (runs all steps to build image for production and calls for docker compose)'

```Makefile
# Target to run the production step
deploy: prod_step compose

# Target to build the Docker image for the production step
prod_step:
	docker build --target production -t andromeda-server .

# Target to run Docker Compose
compose:
	docker compose --profile server up -d
```

2. See error
```
5.586 src/infra/auth/user-roles-decorator.ts:2:10 - error TS2305: Module '"@prisma/client"' has no exported member 'UserRole'.
5.586 
5.586 2 import { UserRole } from '@prisma/client'
5.586            ~~~~~~~~
5.586 src/infra/database/prisma/mappers/prisma-users-mapper.ts:3:18 - error TS2305: Module '"@prisma/client"' has no exported member 'User'.
5.586 
5.586 3 import { Prisma, User as PrismaUser } from '@prisma/client'
5.586                    ~~~~
5.586 src/infra/database/prisma/mappers/prisma-users-mapper.ts:19:39 - error TS2694: Namespace '"/usr/src/app/node_modules/.prisma/client/default".Prisma' has no exported member 'UserUncheckedCreateInput'.
5.586 
5.586 19   static toPrisma(user: User): Prisma.UserUncheckedCreateInput {
5.586                                          ~~~~~~~~~~~~~~~~~~~~~~~~
5.586 src/infra/http/controllers/create-account.controller.ts:9:10 - error TS2305: Module '"@prisma/client"' has no exported member 'UserRole'.
5.586 
5.586 9 import { UserRole } from '@prisma/client'
5.586            ~~~~~~~~
5.586 
5.587 Found 8 error(s).
5.587 
5.625 npm notice 
5.625 npm notice New minor version of npm available! 10.2.4 -> 10.5.0
5.626 npm notice Changelog: <https://github.com/npm/cli/releases/tag/v10.5.0>
5.626 npm notice Run `npm install -g npm@10.5.0` to update!
5.626 npm notice 
------
Dockerfile:43
--------------------
  41 |     COPY --chown=node:node . .
  42 |     
  43 | >>> RUN ["npm", "run", "build"]
  44 |     
  45 |     RUN npm ci --only=production --omit=dev && npm cache clean --force
--------------------
ERROR: failed to solve: process "npm run build" did not complete successfully: exit code: 1

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/u5fv1s29wzctx6vo61k1mma6f
make: *** [Makefile:32: prod_step] Error 1
```

### Expected behavior (optional)

The build step should not conflict with @prisma/client when running `npm run build` on the Dockerfile build step.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native"]
}

datasource db {
  provider = "postgresql"
  url      = env("POSTGRES_URL")
}

model Company {
  id          String    @id @default(uuid())
  name        String
  description String
  address     String
  city        String
  state       String
  country     String
  createdAt   DateTime  @default(now()) @map("created_at")
  updatedAt   DateTime? @updatedAt @map("updated_at")

  businessUnits BusinessUnit[]
  members       User[]

  @@map("companies")
}

model BusinessUnit {
  id          String    @id @default(uuid())
  companyId   String    @map("company_id")
  name        String
  description String
  address     String
  city        String
  state       String
  createdAt   DateTime  @default(now()) @map("created_at")
  updatedAt   DateTime? @updatedAt @map("updated_at")

  company Company @relation(fields: [companyId], references: [id])

  members User[]

  @@map("business_units")
}

enum UserRole {
  MANAGER
  ADMINISTRATOR
  OPERATOR
}

model User {
  id             String     @id @default(uuid())
  companyId      String     @map("company_id")
  businessUnitId String?    @map("business_unit_id")
  name           String
  email          String     @unique
  password       String
  roles          UserRole[] @default([OPERATOR])
  createdAt      DateTime   @default(now()) @map("created_at")
  updatedAt      DateTime?  @updatedAt @map("updated_at")

  businessUnit BusinessUnit? @relation(fields: [businessUnitId], references: [id])
  Company      Company       @relation(fields: [companyId], references: [id])

  @@map("users")
}
```

- OS: Linux Ubuntu 22.04
- Database: Postgres
- Node.js version: 21-alpine

**Comment by ludralph** [comment]
Hi @dufrtss 👋 

Thank you for raising this question.

The error messages indicate missing types like `UserRole` and `User` from `@prisma/client`, suggesting that the Prisma client hasn't been generated yet.

Can you `RUN prisma generate` command before the RUN ["npm", "run", "build"] command in the `BUILD FOR PRODUCTION` stage?

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by dufrtss** [reply]
Hi @ludralph 👋🏻 

Thank you for your reply.

I tried this before but felt like code was getting messier and other problems appeared, here is the error output when I put RUN prisma generate right before the build.

### Input:
```Dockerfile
RUN ["npm", "run", "prisma:generate"]
```

### Output:
```
2024-03-07 13:08:18 prisma:warn We could not find your Prisma schema at `prisma/schema.prisma`.
2024-03-07 13:08:18 If you have a Prisma schema file in a custom path, you will need to run
2024-03-07 13:08:18 `prisma generate --schema=./path/to/your/schema.prisma` to generate Prisma Client.
2024-03-07 13:08:18 If you do not have a Prisma schema file yet, you can ignore this message.
2024-03-07 13:08:18 npm notice New minor version of npm available! 10.2.4 -> 10.5.0
2024-03-07 13:08:18 npm notice Changelog: <https://github.com/npm/cli/releases/tag/v10.5.0>
2024-03-07 13:08:18 npm notice Run `npm install -g npm@10.5.0` to update!
2024-03-07 13:08:18 npm notice
2024-03-07 13:08:20 
2024-03-07 13:08:20 > andromeda-server@0.0.1 prisma:migrate
2024-03-07 13:08:20 > npx prisma migrate deploy
2024-03-07 13:08:20
2024-03-07 13:08:25 Error: Could not find a schema.prisma file that is required for this command.
2024-03-07 13:08:25 You can either provide it with --schema, set it as `prisma.schema` in your package.json or put it into the default location ./prisma/schema.prisma https://pris.ly/d/prisma-schema-location
```

If I copy prisma to my production step to make the build recognize the schema and migrations inside it, like this:

### Input:
```Dockerfile
COPY --chown=node:node --from=build /usr/src/app/primsma ./prisma
```

I get this error.

### Output:
```
2024-03-07 13:16:35 sh: nest: not found
```

## Conclusions
I think this is making the build too complicated and it feels wrong, for instance, inside my production build, it would be preferred to have only the dist folder and node_modules, I believe, adding more dependencies such as the prisma folder and now installing the nest cli inside the container, is it really necessary for production?

**Reply by ludralph** [reply]
Hi @dufrtss 

Would a simpler docker file like this work for you?
```
# Build stage
FROM node:18-alpine AS builder

WORKDIR /app

COPY package*.json ./
RUN npm ci

COPY . .

# Generate Prisma client
RUN npx prisma generate

# Build your application
RUN npm run build

# Production stage
FROM node:18-alpine

WORKDIR /app

COPY --from=builder /app/dist ./dist
COPY --from=builder /app/node_modules/.prisma ./node_modules/.prisma
COPY --from=builder /app/node_modules/@prisma ./node_modules/@prisma
COPY package*.json ./

RUN npm ci --only=production

# Copy only the necessary Prisma files
COPY --from=builder /app/prisma/schema.prisma ./prisma/

CMD ["node", "dist/main"]
```
This Dockerfile does the following:
1. In the build stage:
    - It installs all dependencies, including dev dependencies.
    - Generates the Prisma client.
    - Builds your application.
2. In the production stage:
    - It copies only the built application, the Prisma client, and production dependencies.
    - It copies only the necessary Prisma files (in this case, just the schema).

**Comment by HakamRaza** [comment, marked as answer]
For my project, it seems to work fine by using the following structure. Only 2 folder needed (`dist, node_modules`) and 2 files (`package.json and package-lock.json`)

```Dockerfile
######### First Stage #############

FROM node:20-alpine AS development

WORKDIR /app

COPY --chown=node:node ./app ./

RUN npm ci \
    && npx prisma generate \
    && npm run build \
    && npm prune --omit=dev


######### Second Stage #############

FROM node:20-alpine AS production

WORKDIR /var/www/app

# Create the necessary directories with correct permissions
COPY --chown=node:node --from=development /app/node_modules ./node_modules
COPY --chown=node:node --from=development /app/dist ./dist
COPY --chown=node:node --from=development /app/package*.json ./

# Switch to the non-root user
USER node

EXPOSE 3000

CMD ["node", "dist/src/main"]
```
where my project folder has the following structure:

```
/myproject
  Dockerfile
  /app
    package.json
    ...
    ... (other nestjs files)
    ...
    /src
       /prisma
```

package.json definition:
```
....
  "prisma": {
    "schema": "./src/prisma/schema.prisma"
  }
....
```
Hope it helps.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;

**Comment by aisartag** [comment]
is seem to work


######### First Stage #############
FROM node:22-alpine3.21 AS base


ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"

RUN corepack enable&& corepack install -g pnpm@latest-10

FROM base AS build

WORKDIR /app

COPY . .

RUN pnpm install --frozen-lockfile
RUN pnpm run build 



######### Second Stage #############
FROM base AS prod
WORKDIR /app
COPY . .
RUN pnpm install --frozen-lockfile --prod
RUN pnpm add -D prisma
RUN pnpx prisma generate 


######### Third Stage #############
FROM base AS final

ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"

RUN corepack enable&& corepack install -g pnpm@latest-10

WORKDIR /var/www/app

# Create the necessary directories & files
COPY --from=prod /app/node_modules ./node_modules
COPY --from=prod /app/prisma ./prisma
COPY --from=build /app/package.json ./
COPY --from=build /app/dist ./dist

CMD ["pnpm", "run", "start:migrate:prod"]


in package.json
 "scripts": {
    .....
    "start:migrate:prod": "pnpm prisma migrate deploy && node dist/main",

it works but I'm new to docker and corrections are gladly accepted
